tests/framework/prototype.py

```diff
@@ -0,0 +1,245 @@
+#          specifies a field that wants to mutate with multiple candidate values.
+# Example: "SimpleMutator"
+class Mutator:
+    def __init__(self, baseCR, deltaSets: list[DeltaSet]):
+        self.baseCR = baseCR
+        self.deltaSets = deltaSets
+    # You need to define your mutate() function by inheriting `Mutator`. It should return a new CR.
+    def mutate(self):
+        pass
+    # Apply delta to the base (custom resource).
+    def apply_delta(self, base, delta, steps):
+        root = copy.deepcopy(base)
+        curr = search_path(root, steps[:-1])
+        curr[steps[-1]] = delta
+        return root
+
+# Rule: Rule is used to check whether the actual cluster state is the same as our expectation after a CREvent.
+#       We can infer the expected state by CR YAML file, and get the actual cluster state by Kubernetes API.
+# Example: "HeadPodNameRule"
+class Rule:
+    def __init__(self):
+        pass
+    # The rule will only be checked when `trigger_condition` is true. For example, we will only check
+    # "HeadPodNameRule" when "spec.headGroupSpec" is defined in CR YAML file.
+    def trigger_condition(self, cr=None) -> bool:
+        return True
+    def assertRule(self, cr=None, namespace='default'):
+        raise NotImplementedError
+
+# RuleSet: A set of Rule
+class RuleSet:
+    def __init__(self, rules: list[Rule]):
+        self.rules = rules
+    def checkRuleSet(self, cr, namespace):
+        for rule in self.rules:
+            if rule.trigger_condition(cr):
+                rule.assertRule(cr, namespace)
+
+# CREvent: Custom Resource Event can be mainly divided into 3 categories.
+#   (1) Add (create) CR (2) Update CR (3) Delete CR
+#
+# The member functions integrate together in `trigger()`.
+#   [Step1] exec(): Execute a command to trigger the CREvent. For example, create a CR by a
+#                  `kubectl apply` command.
+#   [Step2] wait(): Wait for the system to converge.
+#   [Step3] checkRuleSets(): When the system converges, check all registered RuleSets.
+class CREvent:
+    def __init__(self, cr, ruleSets: list[RuleSet], timeout, namespace):
+        self.ruleSets = ruleSets
+        self.timeout = timeout
+        self.namespace = namespace
+        self.cr = cr
+    def trigger(self):
+        self.exec()
+        self.wait()
+        self.checkRuleSets()
+    def exec(self):
+        raise NotImplementedError
+    def wait(self):
+        time.sleep(self.timeout)
+    def checkRuleSets(self):
+        for rs in self.ruleSets:
+            rs.checkRuleSet(self.cr, self.namespace)
+
+'''
+My implementations
+'''
+class SimpleMutator(Mutator):
+    def mutate(self):
+        for deltaSet in self.deltaSets:
+            steps = deltaSet.find_path(self.baseCR)
+            if steps:
+                for delta in deltaSet.iterate():
+                    yield self.apply_delta(self.baseCR, delta, steps)
+            else:
+                print("The path does not exist")
```

Comment:
Should this lead to a test failure?

Reply:
I removed the function `apply_delta`, and used `JsonPatch.apply()` instead.